ngraph/frontend/frontend_manager/include/frontend_manager/frontend_manager.hpp

```diff
@@ -0,0 +1,609 @@
+
+            /// \brief Modifies the graph to use new inputs instead of existing ones. New inputs
+            /// should completely satisfy all existing outputs.
+            /// \param inputs Array of new input places
+            virtual void override_all_inputs(const std::vector<Place::Ptr>& inputs);
+
+            /// \brief Leaves only subgraph that are defined by new inputs and new outputs.
+            /// \param inputs Array of new input places
+            /// \param outputs Array of new output places
+            virtual void extract_subgraph(const std::vector<Place::Ptr>& inputs,
+                                          const std::vector<Place::Ptr>& outputs);
+
+            ///// Setting tensor properties  /////
+
+            /// \brief Defines all possible shape that may be used for this place; place should be
+            /// uniquely refer to some data. This partial shape will be converted to corresponding
+            /// shape of results ngraph nodes and will define shape inference when the model is
+            /// converted to ngraph.
+            /// \param place Model place
+            /// \param shape Partial shape for this place
+            virtual void set_partial_shape(Place::Ptr place, const ngraph::PartialShape& shape);
+
+            /// \brief Returns current partial shape used for this place
+            /// \param place Model place
+            /// \return Partial shape for this place
+            virtual ngraph::PartialShape get_partial_shape(Place::Ptr place) const;
+
+            /// \brief Sets new element type for a place
+            /// \param place Model place
+            /// \param type New element type
+            virtual void set_element_type(Place::Ptr place, const ngraph::element::Type& type);
+
+            /// \brief Freezes a tensor with statically defined value or replace existing value for
+            /// already constant node or tensor
+            /// \param place Tensor place
+            /// \param value Value for tensor place representing a memory buffer
+            virtual void set_tensor_value(Place::Ptr place, const void* value);
+
+            /// \brief Defines partial value (lower bound and upper bound) for a tensor place
+            /// TODO: more details for minValue and maxValue format; who defines shape?
+            /// \param place Tensor place
+            /// \param minValue Lower bound of partial value for tensor place
+            /// \param maxValue Upper bound of partial value for tensor place
+            virtual void set_tensor_partial_value(Place::Ptr place,
+                                                  const void* minValue,
+                                                  const void* maxValue);
+        };
+
+        /// \brief An interface for identifying a frontend for a particular framework.
+        /// Provides an ability to load and convert of input model
+        class FRONTEND_API FrontEnd
```

Comment:
In this file we have several classes, can we create separate files for such classes?

General comment/question:

@slyalin @nosovmik At the current moment this API looks really difficult and overloaded for me. I don't understand what and how I should do in order to implement frontend.
For example, if I want to implement my reader what should I do? What class should I reimplement? I hope that it is a `FrontEnd`. I see a lot of different methods in this class, for example:
```
           virtual InputModel::Ptr load_from_file(const std::string& path) const;
            virtual InputModel::Ptr load_from_files(const std::vector<std::string>& paths) const;
            virtual InputModel::Ptr load_from_memory(const void* model) const;
            virtual InputModel::Ptr load_from_memory_fragments(const std::vector<const void*>& modelParts) const;
            ...
```
What methods should I implement? All of these are needed or not...
What about other structures? Should I use something?

Moreover all of this functionality is used now? Or if I understand right we have a big part of this API which won't be used. Can we extend such API in future when we will have necessity?

As I understood not only me don't understand the full set of functionality in this PR.

Maybe I am wrong, but I cannot understand for me the full API from this PR, and I want to be sure that the API doesn't have any redundant methods or objects.

Reply:
Agree. 

Looks strange that we skipped 3rd session about frontends review on arch forum.. Which was supposed to be API-wise dedicated

Reply:
We iterate over it multiple times. Yes, in small groups, with people who is going to implement this API, because the entire API is too big to be reviewed with all folks, because it consist of many methods that are specialized for MO usages. But initial list of services (again, not in a form of list of methods, but in the list of services and properties that we age going to obtain from the API) were reviewed. 4-hours video exists with description of all 4 classes.

Specifically for those load methods -- we discussed it multiple times already, that they are on reconstruction right now. But even in the current state (not ideal, though) they are designed to load model regardless of where it is represented -- in file, in memory or deserialized from stream. The specific structure of the model (number of files possible, in-memory representation) is completely FW dependent and cannot be specified here.

Reply:
As it is a kind of internal/advanced API that is not intended to be used by normal users, we are merging it to master to gather more feedback and unlock other FEs development (to, again, gather more feedback). Regardless of our feedback to particular methods, I don't think that the entire design will change.

Reply:
Splitting to separate classes (frontend_manager, frontend, input_model, place) is done